e2e/test/iothub/service/MessagingClientE2ETests.cs

```diff
@@ -232,5 +235,60 @@ public async Task MessageIdDefaultSetToGuid_SendEventSetMessageIdIfNotSet()
             messageWithoutId.MessageId.Should().NotBeNullOrEmpty();
             messageWithId.MessageId.Should().Be(messageId);
         }
+
+        [LoggedTestMethod, Timeout(TestTimeoutMilliseconds)]
+        [DataRow(IotHubTransportProtocol.Tcp)]
+        [DataRow(IotHubTransportProtocol.WebSocket)]
+        public async Task MessagingClient_SendToNonexistentDevice_ThrowIotHubServiceException(IotHubTransportProtocol protocol)
+        {
+            // arrange
+            var options = new IotHubServiceClientOptions
+            {
+                Protocol = protocol
+            };
+            using var sender = new IotHubServiceClient(TestConfiguration.IoTHub.ConnectionString, options);
+            await sender.Messaging.OpenAsync().ConfigureAwait(false);
+
+            // act
+            var message = new Message(new byte[10]); // arbitrary payload since it shouldn't matter
+            Func<Task> act = async () => await sender.Messaging.SendAsync("nonexistent-device-id", message).ConfigureAwait(false);
+
+            // assert
+            var error = await act.Should().ThrowAsync<IotHubServiceException>();
+            error.And.StatusCode.Should().Be(HttpStatusCode.NotFound);
+            error.And.ErrorCode.Should().Be(IotHubErrorCode.DeviceNotFound);
+
+            await sender.Messaging.CloseAsync().ConfigureAwait(false);
```

Comment:
Perhaps this should be in a finally statement, because otherwise it won't run if the asserts fire.